src/RateLimiter/CounterInterface.php

```diff
@@ -0,0 +1,23 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Yii\Web\RateLimiter;
+
+/**
+ * CounterInterface describes rate limiter counter
+ */
+interface CounterInterface
```

Comment:
I suggests

```php
interface CounterInterface
{
  public function id(string $id): self;
  // required for all algorithm of counters (GCRA, counter based...) to gnerate statistic
  public function limit(int $limit): self;    
  // required for all algorithm of counters (GCRA, counter based...) to gnerate statistic
  public function interval(int $interval): self;
  public function incrementAndGetStatistic(): CounterStatistic;
  // required for real statistic after the one or more middleware is handled
  public function getStatistic(): CounterStatistic;
}
```

Reply:
I don't think expansion of the interface is needed.

1. limit and interval are counter configuration details. These aren't used by middleware or any client so it's better to leave these as constructor arguments.
2. There's currently no client for `getStatistic()`. What's the use case?

Reply:
```php
    public function process(ServerRequestInterface $request, RequestHandlerInterface $handler): ResponseInterface
    {
        $this->counter->setId($this->generateId($request));
        $result = $this->counter->incrementAndGetResult();
        if ($result->isLimitReached()) {
            $response = $this->createErrorResponse();
        } else {
            $response = $handler->handle($request);  // <--- if handle is run long time, then $result is totally outdated
        }
        return $this->addHeaders($response, $result);
    }
```

If handle is run long time, then $result is totally outdated, therefore, a method is needed that get the current state. (`getResult === getStatistic`)

![rate limiter - Google Táblázatok](https://user-images.githubusercontent.com/9432032/72528458-595acb00-386b-11ea-8d06-a7a9a4373947.png)

Reply:
Then incrementAndGetStatistic() result isn't used. Do you propose to separate increment and getting result?

Reply:
if limit is reached, then result is usable, because in that case he immediately returns. I do not think it can be separated, because in that case increment loses the possibility of atomic operation.

Reply:
If a separated, then in this case it is easy for both requests to reach the limit, even though one could have run.

![rate limiter - Google Táblázatok](https://user-images.githubusercontent.com/9432032/72532079-05ec7b00-3873-11ea-982e-764298d06328.png)

Reply:
So you want this?

```php
public function process(ServerRequestInterface $request, RequestHandlerInterface $handler): ResponseInterface
{
    $this->counter->setId($this->generateId($request));
    $result = $this->counter->incrementAndGetResult();

    if ($result->isLimitReached()) {
        $response = $this->createErrorResponse();
    } else {
        $response = $handler->handle($request);
        $result = $this->counter->getResult();
    }

    return $this->addHeaders($response, $result);
}
```

Reply:
Yes

Reply:
In this case we'll have to:

1. Store last increment time.
2. Do two cache reads instead of one in case of not rate-limited request.

Reply:
I'm not sure it worth it. Yes, headers may get outdated.

In the worst case client will be mis-informed about rate allowance and will make extra request that would be responded with "429 Too Many Requests". I don't see it as a big deal.

Same is likely to happen because of networking reasons: if internet connection is slow or shaky.

Reply:
> Do two cache reads instead of one in case of not rate-limited request.

depend on implementation.

> In the worst case client will be mis-informed about rate allowance and will make extra request that will be responded with "429 Too Many Requests". I don't see it as a big deal.

I don't think the interface should be designed less functionality. If you feel it is not necessary, you may want to design the middleware so that it is not capable.

Reply:
> limit and interval are counter configuration details. These aren't used by middleware or any client so it's better to leave these as constructor arguments.

The current design is not logical. The resulting objects have minimal reusability.

current implementation:
```php
$counter1 = new CacheCounter(10,100, $cache);
$middleware1 = new RateLimiter($counter1, ...);
$counter2 = new CacheCounter(100,100, $cache);
$middleware2 = new RateLimiter($counter2, ...);
```

instead of
```php
$counter=new Counter($cache);
$middleware1= new RateLimiter($counter, 10, 100, ...);
$middleware2= new RateLimiter($counter, 100, 100, ...);
// or extended implementacion
$middlewareMulti = new RateLimiter($counter, ...)
->withAddLimit(10,100)
->withAddLimit(5000, 3600);
```

Reply:
> depend on implementation.

You have to do a second read if you want fresh information. There's no way you can avoid that.

> I don't think the interface should be designed less functionality. If you feel it is not necessary, you may want to design the middleware so that it is not capable.

More functionality means more complicated use and more complicated implementation. Less functionality is a good thing (if that's enough functionality to solve the problem).

Reply:
I don't get your last example :(

Reply:
> You have to do a second read if you want fresh information. There's no way you can avoid that.

simple example caching last result of increment.

> More functionality means more complicated use and more complicated implementation. Less functionality is a good thing (if that's enough functionality to solve the problem).

I don't think it would complicate the process, since the result in `incrementAndGetResult` should be able to be generated anyway.
However, `getResult` itself is also suitable for processes such as having the administrator list the problem rate limiter in the admin UI, eg. supports.

Reply:
> simple example: store inside of object the last result of increment.

But then you're not solving anything because information will be equal to the one you get at the time of increment.

Reply:
> But then you're not solving anything because information will be equal to the one you get at the time of increment.

I'm just saying that +1 reading is avoided, and even middleware can be implemented without using it, but depriving the interface is not good.

Reply:
@yiisoft/reviewers need more opinions on this issue.

Reply:
> I don't get your last example :(

As I see @kamarton suggests to clone counter inside limiter. Am I right?
In my opinion it is a bad decision: user may want to get the exact instance of counter he gave to a limiter and to get its values, serialize it or do anything else. So I think it's better to store interval and limit in counter constructor.

---------------
A bit about `getStatistic`.
In my opinion percentage of cases when this approach is reasonable is too low. @kamarton, can you name some cases when it is critical to handle the situation with different results during one request?

Reply:
For now I've decided to merge pull request as is. @kamarton if you want to discuss it further, please create new issue or a forum topic. Thanks.